File: parser.py
"""
Static AST parser for brittle Playwright / Selenium locators.

Walks a Python source file and extracts locator strings plus the
interaction that follows them (click, fill, etc.) so the translator
and generator can migrate each call into AgentQL.
"""

from __future__ import annotations

import ast
import re
from dataclasses import dataclass
from enum import Enum
from pathlib import Path
from typing import Sequence


class LocatorKind(str, Enum):
    """How the locator string was expressed in the source."""

    CSS = "css"
    XPATH = "xpath"
    UNKNOWN = "unknown"


class Interaction(str, Enum):
    """Downstream action applied to the located element."""

    CLICK = "click"
    FILL = "fill"
    TYPE = "type"
    CHECK = "check"
    SELECT = "select_option"
    HOVER = "hover"
    WAIT = "wait_for"
    OTHER = "other"
    NONE = "none"


_INTERACTION_MAP: dict[str, Interaction] = {
    "click": Interaction.CLICK,
    "fill": Interaction.FILL,
    "type": Interaction.TYPE,
    "check": Interaction.CHECK,
    "select_option": Interaction.SELECT,
    "hover": Interaction.HOVER,
    "wait_for": Interaction.WAIT,
}

_PAGE_DIRECT_ACTIONS = frozenset(_INTERACTION_MAP)
_CHAIN_ACTIONS = frozenset(_INTERACTION_MAP)
_LOCATOR_BUILDERS = frozenset({"locator", "query_selector", "query_selector_all"})
_FIND_ELEMENT_METHODS = frozenset({"find_element", "find_elements"})

_XPATH_PREFIX = re.compile(r"^\s*(//|\(\s*\.?/|\./)")
@dataclass(frozen=True, slots=True)
class ParsedLocator:
    """One locator occurrence discovered in a legacy script.

    Attributes:
        selector: Raw CSS/XPath string from the call site.
        kind: Heuristic classification of *selector*.
        interaction: Action chained or applied to the locator.
        lineno: 1-based line of the locator builder / call.
        col_offset: 0-based column of that node.
        fill_value: Literal value for fill/type when available.
        raw_call: Original method name (e.g. ``fill``, ``find_element``).
    """

    selector: str
    kind: LocatorKind
    interaction: Interaction
    lineno: int
    col_offset: int
    fill_value: str | None = None
    raw_call: str | None = None


def classify_selector(selector: str) -> LocatorKind:
    """Best-effort CSS vs XPath classification for a selector string."""
    if not selector or not selector.strip():
        return LocatorKind.UNKNOWN

    stripped = selector.lstrip()
    if stripped.startswith("xpath=") or _XPATH_PREFIX.match(selector):
        return LocatorKind.XPATH
    # Bare absolute path is uncommon for CSS; treat as XPath-ish.
    if stripped.startswith("/") and not stripped.startswith("//"):
        return LocatorKind.XPATH
    return LocatorKind.CSS


def _literal_str(node: ast.AST | None) -> str | None:
    """Return the string value of a constant node, else ``None``."""
    if isinstance(node, ast.Constant) and isinstance(node.value, str):
        return node.value
    return None


def _append_locator(
    bucket: list[ParsedLocator],
    *,
    selector: str,
    kind: LocatorKind,
    interaction: Interaction,
    lineno: int,
    col_offset: int,
    fill_value: str | None,
    raw_call: str,
) -> None:
    """Construct and store a :class:`ParsedLocator`."""
    bucket.append(
        ParsedLocator(
            selector=selector,
            kind=kind,
            interaction=interaction,
            lineno=lineno,
            col_offset=col_offset,
            fill_value=fill_value,
            raw_call=raw_call,
        )
    )


class LocatorVisitor(ast.NodeVisitor):
    """Collect Playwright/Selenium locator usages from an AST.

    Supported shapes:
      * ``page.locator(sel).click()`` / ``.fill(value)`` (and siblings)
      * ``page.click(sel)`` / ``page.fill(sel, value)``
      * ``driver.find_element(By.CSS_SELECTOR, sel)``
    """

    def __init__(self) -> None:
        self.locators: list[ParsedLocator] = []

    def visit_Call(self, node: ast.Call) -> None:
        handled = (
            self._try_chained_locator(node)
            or self._try_page_direct_action(node)
            or self._try_find_element(node)
        )
        if not handled:
            # Descend so nested calls inside arguments are still inspected.
            self.generic_visit(node)

    def _try_chained_locator(self, node: ast.Call) -> bool:
        """Match ``page.locator(...).click()`` / ``.fill()`` style chains."""
        if not isinstance(node.func, ast.Attribute):
            return False
        action_name = node.func.attr
        if action_name not in _CHAIN_ACTIONS:
            return False

        receiver = node.func.value
        if not isinstance(receiver, ast.Call) or not isinstance(receiver.func, ast.Attribute):
            return False
        if receiver.func.attr not in _LOCATOR_BUILDERS or not receiver.args:
            return False

        selector = _literal_str(receiver.args[0])
        if selector is None:
            return False

        interaction = _INTERACTION_MAP[action_name]
        fill_value: str | None = None
        if interaction in {Interaction.FILL, Interaction.TYPE} and node.args:
            fill_value = _literal_str(node.args[0])

        _append_locator(
            self.locators,
            selector=selector,
            kind=classify_selector(selector),
            interaction=interaction,
            lineno=receiver.lineno,
            col_offset=receiver.col_offset,
            fill_value=fill_value,
            raw_call=action_name,
        )
        return True

    def _try_page_direct_action(self, node: ast.Call) -> bool:
        """Match ``page.click(selector)`` / ``page.fill(selector, value)``."""
        if not isinstance(node.func, ast.Attribute):
            return False
        action_name = node.func.attr
        if action_name not in _PAGE_DIRECT_ACTIONS:
            return False
        # Skip chained locator().action() — handled above.
        if isinstance(node.func.value, ast.Call) or not node.args:
            return False

        selector = _literal_str(node.args[0])
        if selector is None:
            return False

        interaction = _INTERACTION_MAP[action_name]
        fill_value: str | None = None
        if interaction in {Interaction.FILL, Interaction.TYPE} and len(node.args) >= 2:
            fill_value = _literal_str(node.args[1])

        _append_locator(
            self.locators,
            selector=selector,
            kind=classify_selector(selector),
            interaction=interaction,
            lineno=node.lineno,
            col_offset=node.col_offset,
            fill_value=fill_value,
            raw_call=action_name,
        )
        return True

    def _try_find_element(self, node: ast.Call) -> bool:
        """Match ``driver.find_element(By.CSS_SELECTOR, '...')`` patterns."""
        if not isinstance(node.func, ast.Attribute):
            return False
        if node.func.attr not in _FIND_ELEMENT_METHODS or len(node.args) < 2:
            return False

        selector = _literal_str(node.args[1])
        if selector is None:
            return False

        kind = self._kind_from_by_strategy(node.args[0], selector)
        _append_locator(
            self.locators,
            selector=selector,
            kind=kind,
            interaction=Interaction.NONE,
            lineno=node.lineno,
            col_offset=node.col_offset,
            fill_value=None,
            raw_call=node.func.attr,
        )
        return True

    @staticmethod
    def _kind_from_by_strategy(strategy: ast.AST, selector: str) -> LocatorKind:
        """Map a Selenium ``By.*`` argument to :class:`LocatorKind`."""
        if isinstance(strategy, ast.Attribute):
            strategy_name = strategy.attr.upper()
            if "XPATH" in strategy_name:
                return LocatorKind.XPATH
            if any(token in strategy_name for token in ("CSS", "ID", "NAME", "CLASS", "TAG")):
                return LocatorKind.CSS
        return classify_selector(selector)


def parse_source(source: str, *, filename: str = "<string>") -> list[ParsedLocator]:
    """Parse *source* and return every locator the visitor recognizes.

    Args:
        source: Python source text containing Playwright/Selenium calls.
        filename: Name attached to AST nodes for error context.

    Returns:
        Locators in source order (stable for deterministic migration).
    """
    tree = ast.parse(source, filename=filename)
    visitor = LocatorVisitor()
    visitor.visit(tree)
    return visitor.locators


def parse_file(path: str | Path) -> list[ParsedLocator]:
    """Read *path* from disk and parse its locators."""
    file_path = Path(path)
    source = file_path.read_text(encoding="utf-8")
    return parse_source(source, filename=str(file_path))


def summarize(locators: Sequence[ParsedLocator]) -> dict[str, int]:
    """Return simple counts keyed by interaction name."""
    counts: dict[str, int] = {}
    for locator in locators:
        key = locator.interaction.value
        counts[key] = counts.get(key, 0) + 1
    return counts

File: web/lib/parser.ts
import type { Interaction, LocatorKind, ParsedLocator } from "./types";

const INTERACTION_MAP: Record<string, Interaction> = {
  click: "click",
  fill: "fill",
  type: "type",
  check: "check",
  select_option: "select_option",
  hover: "hover",
  wait_for: "wait_for",
};

const LOCATOR_BUILDERS = new Set(["locator", "query_selector", "query_selector_all"]);
const FIND_ELEMENT_METHODS = new Set(["find_element", "find_elements"]);
const XPATH_PREFIX = /^\s*(\/\/|\(\s*\.?\/|\.\/)/;

type Token =
  | { kind: "ident"; value: string; index: number }
  | { kind: "string"; value: string; index: number }
  | { kind: "lparen"; index: number }
  | { kind: "rparen"; index: number }
  | { kind: "comma"; index: number }
  | { kind: "dot"; index: number };

export function classifySelector(selector: string): LocatorKind {
  if (!selector || !selector.trim()) return "unknown";
  const stripped = selector.trimStart();
  if (stripped.startsWith("xpath=") || XPATH_PREFIX.test(selector)) return "xpath";
  if (stripped.startsWith("/") && !stripped.startsWith("//")) return "xpath";
  return "css";
}

function posAt(source: string, index: number): { lineno: number; col_offset: number } {
  let lineno = 1;
  let lineStart = 0;
  for (let i = 0; i < index; i++) {
    if (source[i] === "\n") {
      lineno += 1;
      lineStart = i + 1;
    }
  }
  return { lineno, col_offset: index - lineStart };
}

function callStartIndex(source: string, identIndex: number): number {
  let i = identIndex;
  while (i > 0) {
    const prev = i - 1;
    if (source[prev] === ".") {
      i = prev;
      while (i > 0 && /[A-Za-z0-9_]/.test(source[i - 1])) i -= 1;
      continue;
    }
    break;
  }
  return i;
}

function tokenize(source: string): Token[] {
  const tokens: Token[] = [];
  let i = 0;

  const push = (token: Token) => tokens.push(token);

  while (i < source.length) {
    const ch = source[i];

    if (ch === "#") {
      while (i < source.length && source[i] !== "\n") i += 1;
      continue;
    }

    if (ch === " " || ch === "\t" || ch === "\n" || ch === "\r") {
      i += 1;
      continue;
    }

    if (ch === "(") {
      push({ kind: "lparen", index: i });
      i += 1;
      continue;
    }
    if (ch === ")") {
      push({ kind: "rparen", index: i });
      i += 1;
      continue;
    }
    if (ch === ",") {
      push({ kind: "comma", index: i });
      i += 1;
      continue;
    }
    if (ch === ".") {
      push({ kind: "dot", index: i });
      i += 1;
      continue;
    }

    if (ch === "'" || ch === '"') {
      const extracted = readPythonString(source, i);
      if (extracted) {
        push({ kind: "string", value: extracted.value, index: i });
        i = extracted.end;
        continue;
      }
    }

    if (/[A-Za-z_]/.test(ch)) {
      const start = i;
      i += 1;
      while (i < source.length && /[A-Za-z0-9_]/.test(source[i])) i += 1;
      push({ kind: "ident", value: source.slice(start, i), index: start });
      continue;
    }

    i += 1;
  }

  return tokens;
}

function readPythonString(
  source: string,
  start: number,
): { value: string; end: number } | null {
  const quote = source[start];
  if (quote !== "'" && quote !== '"') return null;
  const triple = source.slice(start, start + 3) === quote.repeat(3);
  const delim = triple ? quote.repeat(3) : quote;
  let i = start + delim.length;
  let value = "";

  while (i < source.length) {
    if (source.slice(i, i + delim.length) === delim) {
      return { value, end: i + delim.length };
    }
    if (source[i] === "\\") {
      const next = source[i + 1];
      const escapes: Record<string, string> = {
        n: "\n",
        t: "\t",
        r: "\r",
        "\\": "\\",
        "'": "'",
        '"': '"',
      };
      value += next in escapes ? escapes[next] : (next ?? "");
      i += 2;
      continue;
    }
    value += source[i];
    i += 1;
  }
  return null;
}

function consumeCallArgs(
  tokens: Token[],
  openIndex: number,
): { args: Token[]; end: number } | null {
  if (tokens[openIndex]?.kind !== "lparen") return null;
  const args: Token[] = [];
  let depth = 1;
  let i = openIndex + 1;
  let current: Token | null = null;

  while (i < tokens.length && depth > 0) {
    const tok = tokens[i];
    if (tok.kind === "lparen") {
      depth += 1;
    } else if (tok.kind === "rparen") {
      depth -= 1;
      if (depth === 0) {
        if (current && depth === 0) args.push(current);
        return { args, end: i };
      }
    } else if (tok.kind === "comma" && depth === 1) {
      if (current) args.push(current);
      current = null;
    } else if (depth === 1 && current === null) {
      current = tok;
    }
    i += 1;
  }
  return null;
}

function dottedIdent(tokens: Token[], index: number): string {
  const parts: string[] = [];
  let i = index;
  while (i < tokens.length) {
    const tok = tokens[i];
    if (tok.kind === "ident") {
      parts.push(tok.value);
      const next = tokens[i + 1];
      const after = tokens[i + 2];
      if (next?.kind === "dot" && after?.kind === "ident") {
        i += 2;
        continue;
      }
      break;
    }
    break;
  }
  return parts.join(".");
}

function kindFromByStrategy(strategy: string, selector: string): LocatorKind {
  const name = strategy.toUpperCase();
  if (name.includes("XPATH")) return "xpath";
  if (["CSS", "ID", "NAME", "CLASS", "TAG"].some((token) => name.includes(token))) {
    return "css";
  }
  return classifySelector(selector);
}

export function parseSource(source: string): ParsedLocator[] {
  const tokens = tokenize(source);
  const locators: ParsedLocator[] = [];
  const handledActionAt = new Set<number>();

  for (let i = 0; i < tokens.length; i++) {
    const tok = tokens[i];
    if (tok.kind !== "ident") continue;

    if (LOCATOR_BUILDERS.has(tok.value) && tokens[i + 1]?.kind === "lparen") {
      const call = consumeCallArgs(tokens, i + 1);
      if (!call) continue;
      const selectorTok = call.args[0];
      if (!selectorTok || selectorTok.kind !== "string") continue;

      let j = call.end + 1;
      if (tokens[j]?.kind === "dot") j += 1;
      const actionTok = tokens[j];
      if (!actionTok || actionTok.kind !== "ident") continue;
      const interaction = INTERACTION_MAP[actionTok.value];
      if (!interaction) continue;
      if (tokens[j + 1]?.kind !== "lparen") continue;

      const actionCall = consumeCallArgs(tokens, j + 1);
      let fillValue: string | null = null;
      if (
        actionCall &&
        (interaction === "fill" || interaction === "type") &&
        actionCall.args[0]?.kind === "string"
      ) {
        fillValue = actionCall.args[0].value;
      }

      const start = callStartIndex(source, tok.index);
      const pos = posAt(source, start);
      locators.push({
        selector: selectorTok.value,
        kind: classifySelector(selectorTok.value),
        interaction,
        lineno: pos.lineno,
        col_offset: pos.col_offset,
        fill_value: fillValue,
        raw_call: actionTok.value,
      });
      handledActionAt.add(actionTok.index);
      continue;
    }

    if (FIND_ELEMENT_METHODS.has(tok.value) && tokens[i + 1]?.kind === "lparen") {
      const call = consumeCallArgs(tokens, i + 1);
      if (!call || call.args.length < 2) continue;
      const selectorTok = call.args[1];
      if (selectorTok.kind !== "string") continue;
      const strategyIndex = tokens.findIndex(
        (candidate, idx) => idx > i && candidate === call.args[0],
      );
      const strategy = dottedIdent(tokens, strategyIndex === -1 ? i : strategyIndex);
      const start = callStartIndex(source, tok.index);
      const pos = posAt(source, start);
      locators.push({
        selector: selectorTok.value,
        kind: kindFromByStrategy(strategy, selectorTok.value),
        interaction: "none",
        lineno: pos.lineno,
        col_offset: pos.col_offset,
        fill_value: null,
        raw_call: tok.value,
      });
      continue;
    }

    const interaction = INTERACTION_MAP[tok.value];
    if (!interaction) continue;
    if (handledActionAt.has(tok.index)) continue;
    if (tokens[i + 1]?.kind !== "lparen") continue;
    // Chained locator().action() already handled; page.direct needs an ident receiver.
    const prev = tokens[i - 1];
    if (prev?.kind === "rparen") continue;

    const call = consumeCallArgs(tokens, i + 1);
    if (!call || call.args.length === 0 || call.args[0].kind !== "string") continue;

    let fillValue: string | null = null;
    if (
      (interaction === "fill" || interaction === "type") &&
      call.args[1]?.kind === "string"
    ) {
      fillValue = call.args[1].value;
    }

    const start = callStartIndex(source, tok.index);
    const pos = posAt(source, start);
    locators.push({
      selector: call.args[0].value,
      kind: classifySelector(call.args[0].value),
      interaction,
      lineno: pos.lineno,
      col_offset: pos.col_offset,
      fill_value: fillValue,
      raw_call: tok.value,
    });
  }

  locators.sort((a, b) => a.lineno - b.lineno || a.col_offset - b.col_offset);
  return locators;
}

export function extractGotoUrl(source: string): string | null {
  const tokens = tokenize(source);
  for (let i = 0; i < tokens.length; i++) {
    const tok = tokens[i];
    if (tok.kind !== "ident" || tok.value !== "goto") continue;
    if (tokens[i + 1]?.kind !== "lparen") continue;
    const call = consumeCallArgs(tokens, i + 1);
    if (call?.args[0]?.kind === "string") return call.args[0].value;
  }
  return null;
}
